nats-compiler-plugin/src/main/java/io/ballerina/stdlib/nats/plugin/NatsServiceAnalysisTask.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.ballerina.stdlib.nats.plugin;
+
+import io.ballerina.compiler.api.SemanticModel;
+import io.ballerina.compiler.api.symbols.ModuleSymbol;
+import io.ballerina.compiler.api.symbols.ServiceDeclarationSymbol;
+import io.ballerina.compiler.api.symbols.Symbol;
+import io.ballerina.compiler.api.symbols.TypeDescKind;
+import io.ballerina.compiler.api.symbols.TypeSymbol;
+import io.ballerina.compiler.api.symbols.UnionTypeSymbol;
+import io.ballerina.compiler.syntax.tree.ServiceDeclarationNode;
+import io.ballerina.projects.plugins.AnalysisTask;
+import io.ballerina.projects.plugins.SyntaxNodeAnalysisContext;
+import io.ballerina.stdlib.nats.plugin.PluginConstants.CompilationErrors;
+import io.ballerina.tools.diagnostics.DiagnosticSeverity;
+
+import java.util.List;
+import java.util.Optional;
+
+import static io.ballerina.stdlib.nats.plugin.PluginUtils.validateModuleId;
+
+/**
+ * NATS service compilation analysis task.
+ */
+public class NatsServiceAnalysisTask implements AnalysisTask<SyntaxNodeAnalysisContext> {
+    @Override
+    public void perform(SyntaxNodeAnalysisContext context) {
+        if (!isNatsService(context)) {
+            return;
+        }
+        new NatsServiceValidator(context).validate();
```

Comment:
* Do we need to create `NatsServiceValidator` instances for every service validation. Can't we just use it as follows ?
```java
 public class NatsServiceAnalysisTask implements AnalysisTask<SyntaxNodeAnalysisContext> {
   private final NatsServiceValidator validator;

   public NatsServiceAnalysisTask () {
     this.validator = new NatsServiceValidator();
   }

   @Override
   public void perform(SyntaxNodeAnalysisContext context) {
        if (!isNatsService(context)) {
            return;
        }

        this.validator.validate(context);
        // implement the other logic
   }
 }
````

Reply:
+1 will change

Reply:
@ayeshLK `NatsServiceValidator` is initialized with the `context`. With this change, all the functions in  `NatsServiceValidator` would need the context and other variables initialized in the constructor to be individually passed to them. Do you think that is good design-wise?

Reply:
* @aashikam in here `validate` method in `NatsServiceValidator` is kind of a base method. And as you mention there are few other sub methods available in the `NatsServiceValidator`. 
* But do we need to pass the whole `SyntaxNodeAnalysisContext` to all those sub methods. As I understand those needs only a portion of the original `SyntaxNodeAnalysisContext` object. 
* We could do this breakdown the `SyntaxNodeAnalysisContext` in `validate` method and pass those params into sub methods. 
* So, overall `validate` method will be a method which is a combination of several other methods and it will only have to breakdown the  `SyntaxNodeAnalysisContext` into sub methods.
* Even though you pass `SyntaxNodeAnalysisContext` to constructor you will need to do this context breakdown at some point. I feel that it would be better to do this only in function level.
* You could find a sample [here](https://github.com/ayeshLK/module-ballerina-websub/blob/4ec968242a5d56897b15c66bb470542cf5e2175b/websub-compiler-plugin/src/main/java/io/ballerina/stdlib/websub/task/validator/ServiceDeclarationValidator.java#L58).

Reply:
Will change accordingly. 
On a side note, is the singleton in the sample thread safe?

Reply:
* Yes, since none of the methods are doing any mutation to current state other than updating the ` SyntaxNodeAnalysisContext` it is inherently thread safe (given that update to `SyntaxNodeAnalysisContext` is thread-safe).

Reply:
Fixed in 5d3f69c1720cc0c41b97643cd3c39123fe484b5a